### PR TITLE
CCO-346: Add support for Azure workload identity tokens

### PR DESCRIPTION
**Pull request by dlom**
xref: [CCO-346](https://issues.redhat.com//browse/CCO-346)

* Enhancement proposal: https://github.com/openshift/enhancements/pull/1301
* Corresponding change for `machine-api-operator`: https://github.com/openshift/machine-api-operator/pull/1129
* Additional changes for `machine-api-operator`: https://github.com/openshift/machine-api-operator/pull/1140

Functional changes: 3a304bcc8360c5aa9eac80affdfe2b9f092ad84c
Vendor changes: 15212a72f43ef0ce6f5cd741f8d0217a77d9c0d2

### CODE REVIEW

**Comment by dlom** [comment]
/cc @deads2k 

I've based my work here for the feature gates on the code that you wrote for the cluster ingress operator: https://github.com/openshift/cluster-ingress-operator/pull/908

Could you take a look please and help me with a few questions I had? Specifically, in this operator (and others), there isn't an easy place I can find to pull the [missing version and desired version paramters](https://github.com/openshift/machine-api-provider-azure/commit/79f9958462b03bfa37746ea8831b906f073c0a74#diff-435eecb1d2af40ee96747f88ab71eb2758da989c92f76dcb1f714fc1b300c633R192) that the feature gate checker from library-go wants.  So far I've just filled them in with TODOs and left them unfilled.  

I've also split out the code you wrote into a few functions down at the bottom of main, which I think better encapsulate the intent of the code.  Please let me know what you think!

**Comment by dlom** [comment]
@deads2k I've reworked the featuregating a little bit to comply with the new changes in library-go, and I think I've figured out all of the questions I had before.  Please take a look at my implementation and tell me what you think!

File: cmd/manager/main.go [hunk 1]
```diff
@@ -116,11 +125,33 @@ func main() {
...
+	if err != nil {
+		klog.Fatalf("Failed to create feature gate accessor: %w", err)
+	}
+
+	featureGates, err := awaitEnabledFeatureGates(featureGateAccessor, 1*time.Minute)
+	if err != nil {
+		klog.Fatalf("Failed to get feature gates: %w", err)
+	}
+
+	azureWorkloadIdentityEnabled := featureGates.Enabled("AzureWorkloadIdentity")
```
Comment (JoelSpeed) [inline cmd/manager/main.go:147]: Please use the constant exported in `configv1` in openshift/api

File: pkg/cloud/azure/actuators/machine_scope.go [hunk 2]
```diff
@@ -331,51 +338,114 @@ func updateFromSecret(coreClient controllerclient.Client, scope *MachineScope) e
...
 	}
 
-	subscriptionID, ok := secret.Data[AzureCredsSubscriptionIDKey]
+	subscriptionIDBytes, ok := secret.Data[AzureCredsSubscriptionIDKey]
 	if !ok {
 		return fmt.Errorf("Azure subscription id %v did not contain key %v",
 			secretType.String(), AzureCredsSubscriptionIDKey)
 	}
-	clientID, ok := secret.Data[AzureCredsClientIDKey]
+	clientIDBytes, ok := secret.Data[AzureCredsClientIDKey]
```
Comment (JoelSpeed) [inline pkg/cloud/azure/actuators/machine_scope.go:346]: A lot of the code in this function is duplicated many times, I wonder if extracting it to a helper that first tries the secret and then tries the env var would make this cleaner, WDYT?

**Review by JoelSpeed — COMMENTED** [review]
The termination handler unit tests are failing currently, curious, how do we think these changes will affect the termination handler daemonset? It currently uses the instance metadata to check for a signal that the instance is going away, I suspect it will need to be aware of the changes to how it authenticates there?

**Comment by dlom** [comment]
> how do we think these changes will affect the termination handler daemonset
From what I can tell, the termination handler uses an unauthenticated http GET request to poll for metadata; it doesn't authenticate with azure _at all_, even with the old authentication method.  I'm trying to see if the unit tests are a flake or actually failing because of my change.

**Comment by dlom** [comment]
Turns out the termination handler unit test failures were due to a dependency upgrade deprecating a piece of functionality that the termination handler was using.  Simple fix to swap it out for the intended replacement.

@damdo From what I hear, Joel is out for the next few days; could you please take a look at this PR and potentially give it the stamp of approval? Trying to get this in the merge queue ASAP

**Comment by damdo** [comment]
@dlom 
I'm seeing these oauth error logs during cluster initialization in the E2Es, is this something we expect or is it unrelated?
```
level=error msg=Cluster operator authentication Degraded is True with IngressStateEndpoints_MissingSubsets::OAuthClientsController_SyncError::OAuthServerDeployment_PreconditionNotFulfilled::OAuthServerRouteEndpointAccessibleController_SyncError::OAuthServerServiceEndpointAccessibleController_SyncError::OAuthServerServiceEndpointsEndpointAccessibleController_SyncError::WellKnownReadyController_SyncError: IngressStateEndpointsDegraded: No subsets found for the endpoints of oauth-server
level=error msg=OAuthClientsControllerDegraded: no ingress for host oauth-openshift.apps.ci-op-9hnl2mbc-7c1c7.ci.azure.devcluster.openshift.com in route oauth-openshift in namespace openshift-authentication
level=error msg=OAuthServerDeploymentDegraded: waiting for the oauth-openshift route to contain an admitted ingress: no admitted ingress for route oauth-openshift in namespace openshift-authentication
level=error msg=OAuthServerDeploymentDegraded: 
level=error msg=OAuthServerRouteEndpointAccessibleControllerDegraded: route "openshift-authentication/oauth-openshift": status does not have a valid host address
level=error msg=OAuthServerServiceEndpointAccessibleControllerDegraded: Get "[https://172.30.31.175:443/healthz](https://172.30.31.175/healthz)": dial tcp 172.30.31.175:443: connect: connection refused
level=error msg=OAuthServerServiceEndpointsEndpointAccessibleControllerDegraded: oauth service endpoints are not ready
level=error msg=WellKnownReadyControllerDegraded: failed to get oauth metadata from openshift-config-managed/oauth-openshift ConfigMap: configmap "oauth-openshift" not found (check authentication operator, it is supposed to create this)
level=error msg=Cluster operator authentication Available is False with OAuthServerDeployment_PreconditionNotFulfilled::OAuthServerRouteEndpointAccessibleController_ResourceNotFound::OAuthServerServiceEndpointAccessibleController_EndpointUnavailable::ReadyIngressNodes_NoReadyIngressNodes::WellKnown_NotReady: OAuthServerRouteEndpointAccessibleControllerAvailable: failed to retrieve route from cache: route.route.openshift.io "oauth-openshift" not found
level=error msg=OAuthServerServiceEndpointAccessibleControllerAvailable: Get "[https://172.30.31.175:443/healthz](https://172.30.31.175/healthz)": dial tcp 172.30.31.175:443: connect: connection refused
```

**Review by elmiko — COMMENTED** [review]
this seems like a complicated patch, but i have a question about one of the functions

**Comment by JoelSpeed** [comment]
No feedback from me, LGTM bar Mike's comments. Will need to re-run E2Es anyway since the MAO dependent PR just merged

**Comment by JoelSpeed** [comment]
Looks like we have a permissions issue looking at the logs, https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-api-provider-azure/55/pull-ci-openshift-machine-api-provider-azure-main-e2e-azure-operator/1668432629751353344/artifacts/e2e-azure-operator/gather-extra/artifacts/pods/openshift-machine-api_machine-api-controllers-64cf77bf78-2kh25_machine-controller.log

@dlom Could you please update the permissions required? I think that's an MAO PR

File: pkg/cloud/azure/actuators/machine_scope.go [hunk 3]
```diff
@@ -391,24 +430,38 @@ func updateFromSecret(coreClient controllerclient.Client, scope *MachineScope) e
...
+	value := string(valueBytes)
+	if !ok && strings.TrimSpace(environmentKey) != "" {
+		value = os.Getenv(environmentKey)
+	}
+
+	if strings.TrimSpace(value) == "" {
+		return "", false
+	}
+
+	return value, true
```
[the remark this answers is not in the document]
Comment (newtonheath) [reply pkg/cloud/azure/actuators/machine_scope.go:464]: @dlom - can you address this comment so we can get the merge moving

**Comment by dlom** [comment]
I suspect that now that https://github.com/openshift/machine-api-operator/pull/1140 has merged, the permissions errors will go away :crossed_fingers:

**Comment by dlom** [comment]
Any final comments now that all tests are passing? @JoelSpeed @damdo

**Comment by JoelSpeed** [comment]
/approve

I'm happy with this, will let Damiano add the LGTM

**Review by damdo — COMMENTED** [review]
Thanks @dlom 
/lgtm

**Pull request merged**